packages/react/src/components/Storage/FileUploader/FileUploader.tsx

```diff
@@ -7,279 +7,295 @@ import { Text } from '../../../primitives';
-    const complete = fileStatuses.every((status) => status?.percentage === 100);
-    setSuccess(complete);
 
-    // Creates aggregate percentage to show during downloads
-    const percentage =
-      fileStatuses.reduce((prev, curr) => prev + (curr?.percentage ?? 0), 0) /
-      fileStatuses.length;
-    setPercentage(Math.floor(percentage));
+    return fileStatuses.every((status) => status?.percentage === 100);
+  };
+
+  // Displays if over max files
+  const maxFilesError = fileStatuses.length > maxFiles;
 
+  useEffect(() => {
     // Loading ends when all files are at 100%
     if (Math.floor(percentage) === 100) {
       setLoading(false);
     }
-  }, [fileStatuses]);
+  }, [fileStatuses, percentage]);
 
   useEffect(() => {
     setShowPreviewer(isPreviewerVisible);
   }, [setShowPreviewer, isPreviewerVisible]);
 
-  useEffect(() => {
-    setAllFileNames((allFileNames) => {
-      if (allFileNames.length !== 0) {
-        // only add the new files to allFileNames
-        const diff = files.length - allFileNames.length;
-        const arrayOfNames: string[] = [];
-
-        for (let i = 0; i < diff; i++) {
-          arrayOfNames[i] = files[i].name;
-        }
-
-        return [...arrayOfNames].concat(allFileNames);
-      }
-      return files.map((file) => {
-        return file.name;
-      });
-    });
-    // Check max files size limit if truthy
-    if (maxFiles) {
-      setMaxFilesError(files.length > maxFiles);
-    }
-  }, [files, maxFiles]);
-
   // Previewer Methods
 
-  const progressCallback = (index: number) => {
-    return (progress: { loaded: number; total: number }) => {
-      const percentage = Math.floor((progress.loaded / progress.total) * 100);
-      const status = fileStatusesRef.current[index];
-      fileStatusesRef.current[index] =
-        percentage !== 100
-          ? { ...status, percentage, loading: true }
-          : { ...status, percentage, loading: false, success: true };
-      const addPercentage = [...fileStatusesRef.current];
-      setFileStatuses(addPercentage);
-    };
-  };
-
-  const errorCallback = (index: number) => {
-    return (err: string) => {
-      const status = fileStatusesRef.current[index];
-      fileStatusesRef.current[index] = {
-        ...status,
-        error: true,
-        loading: false,
-        success: false,
-        fileErrors: translate(err.toString()),
+  const progressCallback = useCallback(
+    (index: number) => {
+      return (progress: { loaded: number; total: number }) => {
+        const percentage = Math.floor((progress.loaded / progress.total) * 100);
+        const status = fileStatusesRef.current[index];
+        fileStatusesRef.current[index] =
+          percentage !== 100
+            ? { ...status, percentage, fileState: 'resume' }
+            : { ...status, percentage, fileState: 'success' };
+        const addPercentage = [...fileStatusesRef.current];
+        setFileStatuses(addPercentage);
```

Comment:
Super NIT and not blocking: what do you think about 
```suggestion
        const newFileStatuses = [...fileStatusesRef.current];
        setFileStatuses(newFileStatuses);
```
just to signal these are new file statuses that are getting updated. Same below, and open to disagreement too.

Reply:
fc15c01